packages/fxa-settings/src/lib/oauth/hooks.tsx

```diff
@@ -283,3 +285,37 @@ export function useFinishOAuthFlowHandler(
   }
   return { oAuthDataError: null, finishOAuthFlowHandler };
 }
+
+/**
+ * If we don't have `keyFetchToken` or `unwrapBKey` and it's required, the user needs to
+ * restart the flow. We only store these values in memory, so they don't persist across page
+ * refreshes or browser session restarts. We can't redirect to /signin and reprompt for
+ * password in a browser session restart/crash because the browser stores the flow state in
+ * memory and the query params won't match after a browser session is restored.
+ *
+ * TODO: Can we check session storage for if the user refreshed so we can redirect them
+ * to /signin instead of just showing an error component? FXA-10707
+ */
+export function useOAuthKeysCheck(
+  integration: Pick<Integration, 'type' | 'wantsKeys'>,
+  keyFetchToken?: hexstring,
+  unwrapBKey?: hexstring
+) {
+  if (
+    (isOAuthIntegration(integration) ||
+      isSyncDesktopV3Integration(integration)) &&
+    integration.wantsKeys() &&
+    (!keyFetchToken || !unwrapBKey)
+  ) {
+    return {
```

Comment:
Looking at this fresh today, once I fix tests I'm going to look at returning the error instead of the component/valid status, similar to what we are doing with `oAuthDataError`, so that instead of:

```
  if (!hasValidKeys) {
    return oAuthKeysErrorComponent;
  }
```

We can just do:
```
  if (oAuthKeyCheckError) {
    return <OAuthDataError error={oAuthKeyCheckError} />;
  }